### PR TITLE
Allow source info propagation for null values

### LINKED ISSUE #4647: source info doesn't work when key is undefined
as title, even with `key` defined.

### PR DESCRIPTION
**Pull request by yufeih**
Fixes #4647 , related #4115

### CODE REVIEW

File: src/docfx/lib/json/JsonContractResolver.cs [hunk 1]
```diff
@@ -29,61 +29,69 @@ protected override JsonObjectContract CreateObjectContract(Type objectType)
...
+            return !(value is string) && value is IEnumerable enumerable && !enumerable.GetEnumerator().MoveNext();
+        }
+
+        private static void HandleSourceInfo(JsonProperty property)
+        {
+            if (property.PropertyType.IsGenericType && property.PropertyType.GetGenericTypeDefinition() == typeof(SourceInfo<>))
             {
-                if (!prop.Writable)
+                // Allow source info propagation for null values
+                property.NullValueHandling = NullValueHandling.Include;
```
Comment (OsmondJiang) [inline src/docfx/lib/json/JsonContractResolver.cs]: keep the behavior consistent with normal type null handling: null value means using docfx default value.

File: src/docfx/lib/json/JsonContractResolver.cs [hunk 2]
```diff
@@ -29,61 +29,69 @@ protected override JsonObjectContract CreateObjectContract(Type objectType)
...
+
+                // Do not serialize empty array of null
```
Comment (live1206) [inline src/docfx/lib/json/JsonContractResolver.cs]:
```suggestion
                // Do not serialize empty array or null
```

File: test/docfx.Test/lib/JsonUtilityTest.cs [hunk 3]
```diff
@@ -554,6 +554,37 @@ public void TestJsonMerge(string a, string b, string result)
...
+        }
+
+        [Fact]
+        public void JsonWithTypeErrorsHasCorrectSourceInfo()
+        {
+            var (_, json) = JsonUtility.Parse(
+                "{'next': {'a': ['a1','a2'],'b': 'b'}, 'c': 'c'}".Replace('\'', '"'), "file");
+            var (errors, obj) = JsonUtility.ToObject<ClassWithSourceInfo>(json);
+
+            Assert.Null(obj.Next.A.Value);
```
Comment (yufeih) [inline test/docfx.Test/lib/JsonUtilityTest.cs]: this test showcased that if we simply `return null` without skipping the token in a complex structure, the return object may corrupt. That is why _Report error for invalid yaml header value type as array_ failed when we simply not ignore nulls.

Comment (OsmondJiang) [reply test/docfx.Test/lib/JsonUtilityTest.cs]: No.
With "NullValueHandler = Ignore" This test will pass. I believe IgnoreNullValueHandler did the same thing as you did in this PR.

image

Comment (OsmondJiang) [reply test/docfx.Test/lib/JsonUtilityTest.cs]: @live1206

**Pull request merged**